design/httpproxy-prefix-rewrite.md

```diff
@@ -0,0 +1,262 @@
+  machinery. In this proposal, requests are always fully routed
+  before any changes are made to the URL (this is required by Envoy
+  semantics).
+
+## Background
+
+Request rewriting is ultimately specified as properties of the Envoy [RouteAction](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto#envoy-api-msg-route-routeaction) object.
+The CRD semantics should avoid encouraging uses that can't be supported by Envoy.
+
+The Envoy `prefix_rewrite` configuration is coupled to prefix matching on the route.
+Each `prefix_rewrite` specification needs to be attached to a single prefix match, so each new `prefix_rewrite` needs a new prefix match.
+
+HTTPProxy documents are organized into a collection of trees, where the root is a document that has a `virtualhost` field, and the leaves are documents that have a `routes` field.
+A HTTPProxy tree could be a single document (that is both a root and a leaf), or it could be a series of documents in different namespaces, connected by the `includes` field.
+
+HTTPProxy inclusion semantics allow a single HTTPProxy CRD to be included by more than one root HTTPProxy document.
+This means that multiple routing path prefixes can end up referring to the same HTTPProxy.
+
+To reason about most kinds of request rewriting, the user needs to know the final form of the request.
+In other words, it doesn't make sense to want to rewrite a path prefix unless you have some expectations about the path prefix you will receive.
+This implies that even if the owner of a leaf HTTPProxy doesn't control the root or intermediate documents, they still have fairly concrete expectations around the full URL path to expect.
+
+Although there are a number of ways to conceptualize path prefix rewriting, path regex rewriting requires knowledge of the full request path at the time of writing the regex.
+This is because a regex rewrite can be used to re-order arbitratry portions of the URL path, and this can't be done unless the regex is defined as being applied to the complete path.
+
+## High Level Design
+
+Add a `PathRewrite` field to the HTTPProxy [Route](https://github.com/projectcontour/contour/blob/master/apis/projectcontour/v1/httpproxy.go#L125) object.
+This allows us to specify different kinds of rewrite for URL paths without also needing to specify a specific kind of request rewriting (i.e. we can add host rewrite later).
+
+Locating the `PathRewritePolicy` on the `Route` means that rewriting is controlled by the team that controls the leaf HTTPProxy document.
+This allows a useful initial implementation, but does not prevent adding `PathRewritePolicy` fields higher up the HTTPProxy tree (though allowing that would require detecting and resolving conflicts).
+
+```Go
+type Route struct 
+	...
+
+	// The policy for rewriting the path of the request URL
+	// after the request has been routed to a Service.
+	PathRewrite *PathRewritePolicy `json:"pathRewrite,omitempty"`
```

Comment:
For consistency with the other policies, what about calling this field `rewritePolicy`?

Reply:
If we call it rewrite policy, then that implies that it deals with all kinds of rewrites (e.g host rewrite). The way I'm thinking of dealing with that is a separate rewrite spec, so that's why I've named this to indicate we are specifically dealing with paths. We can add other rewrite kinds in sibling fields.